### Matrices and Vectors In octave

In octave vectors and matrices are written in a similar way, to separate elements of a row we simply leave a space between them, to get to the next row we simply use a semicolon ;. Knowing this we may define row vectors, column vectors and matrices:

In [3]:
displayformat matrix latex; ## Command only works in xeus-octave, and it's meant for pretty printing
v=[0 1 2 3 4] # row vector

error: 'displayformat' undefined near line 1 column 1
v =

   0   1   2   3   4



In [4]:
v=[0 ;1 ;2 ;3; 4] # column vector

v =

   0
   1
   2
   3
   4



In [5]:
m=[0 1 2 3;4 5 6 7; 8 9 10 11;12 13 14 15]# matrix

m =

    0    1    2    3
    4    5    6    7
    8    9   10   11
   12   13   14   15



In octave transposition is done with the ' character, one thing to be aware about is that this also the dagger operation meaning that for complex matrices it also performs conjugation

In [6]:
p=[1 0 0 0; 0 1 0 0; 0 0 0 1; 0 0 1 0]

p =

   1   0   0   0
   0   1   0   0
   0   0   0   1
   0   0   1   0



In [7]:
p'

ans =

   1   0   0   0
   0   1   0   0
   0   0   0   1
   0   0   1   0



Finding eigenvalues, eigenvectors and so on in octave is fairly straightforward

The determinant is obtained with the det function, eigevalues and eigenvectors with eig

In [8]:
det(p)

ans = -1


In [9]:
eig(p) #eigenvalues by default

ans =

  -1
   1
   1
   1



In [20]:
[eivects,eigvals]=eig(p)

eivects =

   0.00000   0.00000   0.00000   1.00000
   0.00000   1.00000   0.00000   0.00000
  -0.70711   0.00000   0.70711   0.00000
   0.70711   0.00000   0.70711   0.00000

eigvals =

Diagonal Matrix

  -1   0   0   0
   0   1   0   0
   0   0   1   0
   0   0   0   1



Eigenvectors come as a matrix so one needs to do  little matrix indexing to recover the actual vectors, In octave one can index elements of a matrix using parenthesis


A(row,column)


in octave : means all elements so for example the element

[insert images]

may be obtained by A(1,1) 

While the first row of A may be obtained by A(:,1)

In [24]:
eivects(:,1)

ans =

   0.00000
   0.00000
  -0.70711
   0.70711



We check that it is indeed an eigenvector

In [25]:
p*eivects(:,1)

ans =

   0.00000
   0.00000
   0.70711
  -0.70711



The upper bound on the error of numerical linear algebra calculations is the so called [condition number](https://en.wikipedia.org/wiki/Condition_number#Matrices) which is defined as the ratio of the maximal and singular values of a matrix, if the condition number is big with respect to 1, then we might expect incorrect results from whatever calculations we do with this matrix

In [30]:
cond(p)

ans =  1


In [31]:
cond(m) # m is a singular matrix so a high condition number is expected

ans =    3.5680e+17


## Linear Systems Basic Gaussian elimination

Consider the system of equations 

\begin{align*}
x+2y+3z=4\\
-2y-4z=6\\
x-y=0
\end{align*}

We may write in [augmented matrix form](https://en.wikipedia.org/wiki/Augmented_matrix) in octave simply by:

In [64]:
B=[1 2 3 4;0 -2 -4 6;1 -1 0 0]

B =

   1   2   3   4
   0  -2  -4   6
   1  -1   0   0



In [65]:
A=B(:,1:3); #we extract the system of equations
b=B(:,4); # and the results from the augmented matrix to use them later

Now we may simply start doing gaussian elimination, the idea is to make all elements but the ones in the diagonal and last row zero, by performing elementary operations on rows (Multiplication, division and addition to other rows), to get some practice on indexing let us do this in octave

In [66]:
B(3,:)=-B(3,:)+B(1,:) #Multiply last row by minus one and add the first row to it

B =

   1   2   3   4
   0  -2  -4   6
   0   3   3   4



In [67]:
B(3,:) =( B(3,:) + (3/2) *B(2,:))/(-3) # Change the third row by adding to it 3/2 times the second row 
# and multiplying the result by -1/3

B =

   1.00000   2.00000   3.00000   4.00000
   0.00000  -2.00000  -4.00000   6.00000
  -0.00000  -0.00000   1.00000  -4.33333



In [68]:
B(2,:)=((1/4)*B(2,:)+B(3,:))*-2 # Change the second row by multiplying it by -1/4 and add the third row to it
# multiply everything by -2 in the end

B =

   1.00000   2.00000   3.00000   4.00000
  -0.00000   1.00000  -0.00000   5.66667
  -0.00000  -0.00000   1.00000  -4.33333



In [69]:
B(1,:)=B(1,:)-3*B(3,:) # Change the first row by adding the last one times -3

B =

    1.00000    2.00000    0.00000   17.00000
   -0.00000    1.00000   -0.00000    5.66667
   -0.00000   -0.00000    1.00000   -4.33333



In [70]:
B(1,:)=B(1,:)-2*B(2,:) # Change the first row by adding the second one times -2

B =

   1.00000   0.00000   0.00000   5.66667
  -0.00000   1.00000  -0.00000   5.66667
  -0.00000  -0.00000   1.00000  -4.33333



By performing Gaussian elimination we found x=5.6667=17/3=y and z=-13/3=-4.33333

#### Using Left Division

In Octave the built in operation for solving linear systems is called Left division, it works on systems of the form $A \vec{x}=\vec{b}$, What left division is equivalent to is $A^{-1} \vec{b}$, and it is written as A\b, so we may solve the same system by this simple command

In [72]:
A\b

ans =

   5.6667
   5.6667
  -4.3333



We indeed see the same result

![System.png](attachment:ba1c1ce6-3f98-4dc5-91d3-b3be3b3cc55d.png)

### **Task 1** Write the system in augmented matrix form, and perform Gaussian elimination, check your answer using left division  

## The LU decomposition

The idea of this matrix decompositio is to write the matrix A as $A=LU$ where U is an upper triangular matrix and L is a lower triangular matrix, How to get one is similar to Gaussian elimination, let us break down the procedure:

[Images]

One May use the LU decomposition to solve determinants, since 

det(A)=det(LU)=det(L)det(U)

and the determinant of a triangular matrix is just the product of the diagonal elements

In [75]:
U=[1 2 3;0 -2 -4;0 0 3]

U =

   1   2   3
   0  -2  -4
   0   0   3



In [76]:
L=[1 0 0;0 1 0;1 1.5 1]

L =

   1.00000   0.00000   0.00000
   0.00000   1.00000   0.00000
   1.00000   1.50000   1.00000



We may find the product of the diagonal elements via a for loop

In [96]:
detL=1;
for i=1:size(L)[1];
    detL*=L(i,i);
end
detL

detL =  1


**Task** Find the determinant of det(A) using the LU decomposition (without using the built in function det()) later compare with the result from det

The other main application of the LU decomposition is solving linear systems. To see how let us follow our example

[image]

As we can see once we have LU decomposition everything reduces to forward and backward substitution

In [97]:
function x = backward(U,b)
% BACKWARD. backward elimination.
% For upper triangular U, x = backward(U,b) solves U*x = b. If an element on Diagonal is zero perform permutations so LU=PA
    n=size(U)(1); # gets size of matrix
    for k = n:-1:1 # go from n to one decreasing
        x(k) = b(k)/U(k,k); # find variable from ax=y
        i = (1:k-1); # range of other variables
        b(i) = b(i) - x(k)*U(i,k);  # subs found variables into b
    end
end

In [100]:
function x = forward(L,x)
% FORWARD. Forward elimination.
% For lower triangular L, x = forward(L,b) solves L*x = b. If an element on Diagonal is zero perform permutations so LU=PA
n= size(L)(1);
for k = 1:n
j = 1:k-1; # same as backward in one step
x(k) = (x(k) - L(k,j)*x(j))/L(k,k);
end
end

In [101]:
forward(L,[4;6;0])

ans =

    4
    6
  -13



In [102]:
backward(U,[4;6;-13])

ans =

   5.6667   5.6667  -4.3333



In [32]:
function [lower,upper]=dolittle(A)
n=size(A)(1);
lower=zeros(n,n);
upper=eye(n,n);
for i=1:n
    for k=i:n
    suma1=0;
    suma2=0;
        for j=1:i
        suma1=suma1+lower(i,j)*upper(j,k);
        suma2=suma2+lower(k,j)*upper(j,i);
        end
    upper(i,k)=A(i,k)-suma1;
    lower(k,i)=(A(k,i)-suma2)/upper(i,i);
    end
end

In [33]:
[l,u]=dolittle(org)

l =    1.0000        0        0
        0   1.0000        0
   1.0000   1.5000   1.0000


u =    1   2   3
   0  -2  -4
   0   0   3


In [35]:
l*u-org

ans =    0   0   0
   0   0   0
   0   0   0


In [72]:
[l, u, p] = lu(org)

l =    1.0000        0        0
   1.0000   1.0000        0
        0   0.6667   1.0000


u =    1   2   3
   0  -3  -3
   0   0  -2


p = Permutation Matrix

   1   0   0
   0   0   1
   0   1   0
